proposals/source-install.md

```diff
@@ -0,0 +1,1133 @@
+# Installing the AWS CLI v2 from source
+
+Proposal         | Metadata
+---------------- | -------------
+**Author**       | Kyle Knapp
+**Status**       | Proposed
+**Created**      | 24-August-2021
+
+## Abstract
+
+This document proposes an official mechanism to install the AWS CLI v2 from
+source along with an official source distribution artifact.
+
+
+## Motivation
+
+The AWS CLI v2 is available as pre-built executables for macOS,
+Linux x86_64, Linux aarch64, and Windows (64-bit). The AWS CLI v2 is also
+available as a Docker image. Generally, these artifacts provide coverage for
+most platforms and environments, but they do not satisfy all use cases:
+
+* The desired platform (e.g, [ARM 32-bit](https://github.com/aws/aws-cli/issues/5426))
+  is not supported by any of the pre-built executables.
+  
+* The environment lacks system dependencies that are required for the pre-built
+  executable. For example, Alpine Linux uses [musl](https://musl.libc.org/),
+  but the current executables require glibc. This causes the pre-built
+  executables to not work out of the box for
+  [Alpine Linux](https://github.com/aws/aws-cli/issues/4685).
+  
+* The environment restricts access to resources only needed by the
+  pre-built executable. For example, on security hardened systems, it does not
+  give permission to shared memory
+  [which is needed](https://github.com/aws/aws-cli/issues/5769)
+  during the bootloading process of the frozen `aws` executable.
+
+* Users want to be able to install the AWS CLI using the package manager
+  standard to their environment (e.g., `brew`, `yum`, `apt`) instead of having
+  to download and use a ZIP, PKG,  MSI, etc. The software available through
+  these package managers are typically maintained by upstream distribution
+  maintainers. When maintainers import software, they  build from source to
+  maintain full control over the code and  packages being imported into their
+  distribution. Therefore, importing a pre-built executable is generally a
+  [non-starter for distro maintainers](https://github.com/aws/aws-cli/issues/4842).
+
+Full support for installing the AWS CLI v2 from source will unblock these use
+cases because users will be able to build and install the AWS CLI v2 for their
+particular environment.
+
+
+### Goals
+
+The approach for installing the AWS CLI v2 from source should satisfy the
+following goals:
+
+1. The source installation process maximizes the number of environments the
+   AWS CLI v2 can be installed on.
+   
+2. The source installation process is straightforward and intuitive. It minimizes
+   the number of cycles required to figure out how to install the AWS CLI from
+   source.
+
+3. The source installation process is language agnostic. While the AWS CLI v2
```

Comment:
This feels like such an odd goal and the result of overthinking and overcomplicating things. The installation process can be language agnostic, but the build process _never_ is. It is the responsibility of users _choosing_ to build from source or install using language-specific tooling to understand how that works. You don't have to *promote* language-specific tooling as an official method and you can guide users to your own packaged and hosted installers, but language-specific methods should be supported and maintained (not "provisional"). As a python project implementing a cli utility, users should be able to simply `pipx install awscliv2`.

If you switch to Go, for example, users will want to be able to use `go install ...` and if this were a Go project that _should_ just work! Switching languages like this would almost certainly qualify as a major release anyway, and so have no backwards combability guarantees, not of the user interface and not of the build process!

Reply:
> It is the responsibility of users *choosing* to build from source or install using language-specific tooling to understand how that works.

What if building from source is their only choice? A large number of the users that will be leveraging this proposal will be the ones that have no other option to install the AWS CLI v2 (i.e., they are running on platforms where there is no official installer). Furthermore, these users are not all going to be experienced in any particular language, and nor should they need to be experienced in order to get a language-agnostic tool installed on their system. For example with v1, using Python tooling was often the only way to get the CLI installed on their system and this often bit non-Python developers whether they had to spend an significant amount of cycles trying to install the CLI or broke other tools installed on their system. I think if there were other valid alternatives to install the AWS CLI v2 for this group of users, the goal could be relaxed/removed.

> language-specific methods should be supported and maintained (not "provisional"). As a python project implementing a cli utility, users should be able to simply `pipx install awscliv2`.

Maybe provisional has the wrong connotations and there's a better word or something needs to be more explicit in the definition? But the idea is that as long as the AWS CLI is a Python project, we will comply to Python standards (e.g., PEP517) to enable building/installing of the AWS CLI using Python tooling. For example with `pipx` (assuming it complies with PEP517 in building a wheel), you should be able to just run `pipx install .` on the repository. In the end, if a specific tool supports PEP517 and only requires support for that standard, we are not going to prevent users from using that specific tool on the CLI codebase; we'd likely just not document/promote using the tool on the codebase.

The idea with labeling it as provisional is that allows the project to adapt to whatever language-tooling is appropriate based on the repository and language community. For example, it does not make sense to support PEP517 if the codebase was entirely written in Go. And yes while if we were to rewrite it in Go, it would likely warrant a major version bump, breaking PEP517 compliance would not necessarily require a complete rewrite. For example:
* We could introduce a new separate executable to the AWS CLI package (similar to the separate  `aws_completer` we ship in the package today) that could be written in a different language.

* A new PEP in the future could replace PEP517 (i.e., come out with a new format to replace wheels). Ideally the CLI project would just support this new PEP while dropping support for PEP517 in order to follow what is recommended for a Python project.

Reply:
>What if building from source is their only choice? A large number of the users that will be leveraging this proposal will be the ones that have no other option to install the AWS CLI v2 (i.e., they are running on platforms where there is no official installer).

This really seems to be taking on an awful lot yourselves and introducing a lot of complexity as a result. At some point, there has to be a boundary between this project and the user's environment. I might suggest focusing on creating official installers for those platforms?

And, the approach detailed here is still building from source and requires a healthy python install to make that work! So I don't really see much difference, I guess. Users that, before, had trouble managing python runtime isolation and installing packages into a specific python runtime will still have trouble, won't they? It is certainly nice to provide the logic for creating the portable and sandbox packages, though. Certainly do appreciate that!

>this often bit non-Python developers whether they had to spend an significant amount of cycles trying to install the CLI or broke other tools installed on their system.

Publishing to PyPi and constraining the recommendation for a python-based cli installation to use `pipx` instead of `pip` might go a long way to reducing these problems, I think.

Reply:
> This really seems to be taking on an awful lot yourselves and introducing a lot of complexity as a result. At some point, there has to be a boundary between this project and the user's environment.

Yeah it's definitely not an insignificant amount of effort to maintain this. The end approach is actually fairly similar to the [bundled installer](https://docs.aws.amazon.com/cli/latest/userguide/install-linux.html#install-linux-bundled) for v1. So we have experience with the approach/boundary and this new approach tries to incorporate lessons learned with that installer. For example, the main differences between the bundled installer and the build process in this proposal are:

* The bundled installer is a zip containing the sdists for all of the AWS CLI v1's dependencies. This was fairly limiting because users could not control what version of a dependency they wanted to install and could not leverage already built wheels (and would have to rebuild the wheel as part of installing the bundled installer). With this build process it just reuses the Python packages already on the system or just downloads them into a virtualenv using pip (which has more flexibility in that it can better choose what artifact to download and install depending on the system).

* The bundled installer had backwards compatibility guarantees that all you need was a Python interpreter and any subsequent upgrade should continue to pass without any new software being installed on the system. This has prevented us from pulling in dependencies that require a compiler or even upgrade some dependencies. With the proposal, there is no backwards compatibility guarantee on needing additional system dependencies.

> I might suggest focusing on creating official installers for those platforms?

Yep we should or even better yet make the AWS CLI v2 installable through standard package managers for those platforms. The primary purpose for this approach is for the development team to stop being the bottleneck for all potential users that currently have no path forward to installing the AWS CLI v2 as it would take a considerable amount of time to reach full coverage for all of the platforms people want to run the AWS CLI. Ideally, in the future, less users would need to use the source distribution, but there is no timetable for when that future would take hold.

> And, the approach detailed here is still building from source and requires a healthy python install to make that work! So I don't really see much difference, I guess. Users that, before, had trouble managing python runtime isolation and installing packages into a specific python runtime will still have trouble, won't they? 

Not necessarily. The idea is that for the user that has no experience in Python nor cares about reusing existing Python dependencies (i.e., just get the AWS CLI v2 installed in the quickest way possible) they should be able to blindly just run the following commands and the build/install just works (because it is abstracting over the common issues such as installing/isolating the packages to a virtualenv):
```
./configure --with-download-deps
make
make install
```
Especially, with v1, non-Python users would just blindly run `pip install awscli` because they would not know any better and not realize there is best practices to installing Python packages.

I do think the Python community has improved in the area of tools enforcing best practices and making them more accessible (e.g., `pipx`), but the concern is these tools are still Python specific and would force non-Python users to install and learn a new tool/terminology outside of their standard package manager or `configure`/`make` workflow.

Reply:
>Not necessarily. The idea is that for the user that has no experience in Python nor cares about reusing existing Python dependencies (i.e., just get the AWS CLI v2 installed in the quickest way possible) they should be able to blindly just run the following commands and the build/install just works (because it is abstracting over the common issues such as installing/isolating the packages to a virtualenv)

Well ok. It seems like a nice idea. But isn't it just trading expertise with python for expertise with make? And probably particular versions of make and the <shell> used (MacOS problems). And some number of other system packages, and the inscrutable errors from make when an assumption in the script/Makefile gets violated. 

There are totally valid use cases where users need to build from source, and providing utilities like this does make it easier (which is awesome!). But it doesn't really seem like this is making anything any easier for the type of user that is going to be having problems managing their system and Python runtimes. This is still a very advanced type of install requiring specialized knowledge and tools. These are orthogonal use cases.

>these tools are still Python specific and would force non-Python users to install and learn a new tool/terminology

When there is no official installer for their platform. Or when their environment restrictions are such that they are prevented from using an official installer (e.g. non-admin user on Windows with awscli v2).

But even then, it feels much lower effort to lean into the packaging tools and practices and expectations for the programming language, and to actively support that as a lowest common denominator (see above, trading one bit of expertise for another). Publishing to PyPi is trivially easy, and addresses a [large swath of the user and developer community](https://pypistats.org/packages/awscli)(~100mil/month!) . Why put so much thought and effort into making things harder for those users?

Reply:
Oh wow, over the last month, botocore and s3transfer are in the top 5, and awscli v1 is #10! People really love these on pypi! https://pypistats.org/top

Reply:
I think this is actually evidence the AWS CLI has a significantly wider userbase than Python users.